### PR TITLE
adding code to ingest indexing pressure stats

### PR DESCRIPTION
**Pull request by turnUpTheChill**
## What does this PR do?
This PR will add functionality to the Metricbeat Elasticsearch `node_stats` module to pull `indexing_pressure` metrics

## Why is it important?
These metrics were added in 7.9 as part of the expansion of the default thread pool, and are thus crucial for reliable monitoring of Elasticsearch. Without them natively available, custom solutions to scrape the endpoint are required for 
something that should be available out of the box. 

## Checklist

- [ ] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## How to test this PR locally

This can be tested locally by building the branch and running `./metricbeat test modules elasticsearch node_stats`

## Related issues

Closes [#27046](https://github.com/elastic/beats/issues/27046)

Relates [#59263](https://github.com/elastic/elasticsearch/issues/59263)

-

### CODE REVIEW

**Comment by Leaf-Lin** [comment]
Hi, 

@elastic/integrations , is it possible for someone to finish this PR which appears to be 99% finished? The indexing pressure is an important metrics for metribeat to track and a useful measure for understanding cluster indexing pressure, please...

File: metricbeat/module/elasticsearch/node_stats/data.go [hunk 1]
```diff
@@ -318,7 +352,7 @@ func eventsMapping(r mb.ReporterV2, m elasticsearch.MetricSetAPI, info elasticse
 				"id":   info.ClusterID,
 			},
 		}
-
+		fmt.Printf("event: %v\n", event)
```
Comment (sayden) [inline metricbeat/module/elasticsearch/node_stats/data.go:355]: Leftover

**Review by sayden — DISMISSED** [review]
None of the fields follow the Elastic naming conventions, please check this guide https://www.elastic.co/guide/en/beats/devguide/current/event-conventions.html

File: metricbeat/module/elasticsearch/_meta/fields.yml [hunk 2]
```diff
@@ -613,7 +613,26 @@
                   type: alias
                 - name: rejected
                   path: elasticsearch.node.stats.thread_pool.write.rejected.count
-                  type: alias
+                  type: alias    
+        - name: indexing_pressure
```
Comment (sayden) [inline metricbeat/module/elasticsearch/_meta/fields.yml:617]: This entire group is not required

File: metricbeat/module/elasticsearch/node_stats/_meta/fields.yml [hunk 3]
```diff
@@ -312,4 +312,21 @@
               type: long
             - name: rejected.count
               type: long
-
+    - name: indexing_pressure
```
Comment (sayden) [inline metricbeat/module/elasticsearch/node_stats/_meta/fields.yml:315]: This entire group must not contain `elasticsearch.node.stats` as a prefix of each field and it must follow naming conventions

File: metricbeat/module/elasticsearch/_meta/fields.yml (hunk 2 above)
Comment (turnUpTheChill) [reply metricbeat/module/elasticsearch/_meta/fields.yml:617]: @sayden Is my understanding correct that the group only has to exist in `node_stats/_meta/fields.yml`?

File: metricbeat/module/elasticsearch/node_stats/_meta/fields.yml (hunk 3 above)
Comment (kaiyan-sheng) [inline metricbeat/module/elasticsearch/node_stats/_meta/fields.yml:315]: Maybe something like this?
```suggestion
    - name: indexing_pressure
       type: group
       fields:
           - name: memory
              type: group
              fields:
                  - name: current.combined_coordinating_and_primary.bytes
                    type: long
                    format: bytes
                    description: xxx
                 - name: total.combined_coordinating_and_primary.bytes
                    type: long
                    format: bytes
                    description: yyy
```

Comment (turnUpTheChill) [reply metricbeat/module/elasticsearch/node_stats/_meta/fields.yml:315]: @kaiyan-sheng thank you! this was very helpful

**Comment by turnUpTheChill** [comment]
@sayden Thank you for your review. I have gone through an made the necessary changes, please let me know if it looks ok. 
cc: @Leaf-Lin

**Comment by kaiyan-sheng** [comment]
@turnUpTheChill Thanks for adding this!! For the failed CLA check, could you make sure your github account links to your elastic email please?

**Comment by kaiyan-sheng** [comment]
cla/check

**Comment by turnUpTheChill** [comment]
cla/check

> @turnUpTheChill Thanks for adding this!! For the failed CLA check, could you make sure your github account links to your elastic email please?

confirmed, but I cannot retrigger the check?

**Review by kuisathaverat — CHANGES_REQUESTED** [review]
This PR has a bunch of new files that should not be new files, I dunno what exactly it is going on, but since the PR has several months I would recommend updating the PR with the changes in the main repo or rebase it.

File: metricbeat/module/elasticsearch/node_stats/_meta/fields.yml [hunk 4]
```diff
@@ -339,3 +339,54 @@
               type: long
             - name: rejected.count
               type: long
+<<<<<<< HEAD
```
Comment (kaiyan-sheng) [inline metricbeat/module/elasticsearch/node_stats/_meta/fields.yml:342]: merge conflict residual

**Comment by sayden** [comment]
@jasonrhodes or @klacabane or @matschaffer can any of you help here, please? I'm a bit overwhelmed with some aws stuff. Thanks!

**Comment by turnUpTheChill** [comment]
> Sounds like this might factor into [elastic/kibana#129351](https://github.com/elastic/kibana/issues/129351) pretty well, I'll dig more into that issue next week once I'm off-call.

@matschaffer they're tangentially related, but this just addresses `indexing_pressure`. I'd tried to address `ingest` pipelines as well, but faced some limitations with the schema definitions never picking up the pipeline metrics properly.

File: metricbeat/module/elasticsearch/node_stats/_meta/fields.yml (hunk 3 above)
Comment (matschaffer) [reply metricbeat/module/elasticsearch/node_stats/_meta/fields.yml:342]: This appears to have been resolved by more recent commits.

**Comment by matschaffer** [comment]
As far as I can tell @turnUpTheChill has already addressed the change requests so far. I've updated it from main to see how the test suite goes. I'll put it on my list to build it and see how the data lands.

**Comment by matschaffer** [comment]
Screen Shot 2022-04-20 at 15 46 02

Seems to be working, we'll need mapping updates for this as well for it to be queryable when running in xpack mode.

But it's graphable via `metricbeat-*` when running `xpack.enabled: false`.

Screen Shot 2022-04-20 at 15 50 35

I'll work on adding it to https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/core/src/main/resources/monitoring-es-mb.json so we can use it in xpack mode as well.

File: metricbeat/module/elasticsearch/node_stats/data.go [hunk 5]
```diff
@@ -257,6 +256,21 @@ var (
...
 			"search": c.Dict("search", threadPoolStatsSchema),
 		}),
+		"indexing_pressure": c.Dict("indexing_pressure", s.Schema{
+			"memory": c.Dict("memory", s.Schema{
+				"current":        c.Dict("current", currentMemoryPressureSchema),
+				"total":          c.Dict("total", totalMemoryPressureSchema),
+				"limit_in_bytes": c.Int("limit_in_bytes"),
+			}),
+		}),
+		"ingest": c.Dict("ingest", s.Schema{
```
Comment (matschaffer) [inline metricbeat/module/elasticsearch/node_stats/data.go:266]: Looks like this still needs fields entries

Comment (turnUpTheChill) [reply metricbeat/module/elasticsearch/node_stats/data.go:266]: @matschaffer as in, these fields need to be entered into the `fields.yml` file?

Comment (matschaffer) [reply metricbeat/module/elasticsearch/node_stats/data.go:266]: yep! the integration tests are currently failing with:

```
Exception: Key 'elasticsearch.node.stats.ingest.total.count' found in event ... is not documented!
```

You should be able to confirm via https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats/detail/PR-28479/20/pipeline

**Comment by klacabane** [comment]
Pushed some fixes to have the tests pass

LGTM

**Review by klacabane — APPROVED** [review]

**Comment by matschaffer** [comment]
Merging given @klacabane 's updates & approval.

**Pull request merged**